pkg/agent/heartbeat.go

```diff
@@ -0,0 +1,89 @@
+package agent
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	timeout  = 3
+	interval = 10000
+
+	online  = "online"
+	offline = "offline"
+
+	service = "service"
+	device  = "device"
+)
+
+// svc keeps info on service live status.
+// Services send heartbeat to nats thus updating last seen.
+// When service doesnt send heartbeat for some time gets marked offline.
+type svc struct {
+	info    Info
+	counter int
+	ticker  *time.Ticker
+	mu      sync.Mutex
+}
+
+type Info struct {
+	Name     string    `json:"name"`
+	LastSeen time.Time `json:"last_seen"`
+	Status   string    `json:"status"`
+	Type     string    `json:"type"`
+	Terminal int       `json:"terminal"`
+}
+
+// Heartbeat specifies api for updating status and keeping track on services
+// that are sending heartbeat to NATS.
+type Heartbeat interface {
+	Update()
+	Info() Info
+}
+
+func NewHeartbeat(name, svctype string) Heartbeat {
+	ticker := time.NewTicker(interval * time.Millisecond)
+	s := svc{
+		info: Info{
+			Name:     name,
+			Status:   online,
+			Type:     svctype,
+			LastSeen: time.Now(),
+		},
+		counter: timeout,
+		ticker:  ticker,
+	}
+	s.listen()
+	return &s
+}
+
+func (s *svc) listen() {
+	go func() {
+		for {
+			select {
+			case <-s.ticker.C:
+				// TODO - we can disable ticker when the status gets OFFLINE
```

Comment:
Can you create an issue for this please?

Reply:
done https://github.com/mainflux/agent/issues/42